goblocks.go

```diff
@@ -5,91 +5,96 @@ import (
 	"fmt"
 	"goblocks/util"
 	"io/ioutil"
+	"log"
 	"os"
 	"os/exec"
+	"path/filepath"
 	"strings"
 )
 
 type configStruct struct {
-	Separator string
+	Separator          string
 	ConfigReloadSignal int //currently unused
-	Actions []map[string]interface{}
+	Actions            []map[string]interface{}
```

Comment:
Why are you using a type-unsafe `map[string]interface{}` here? Can't you use a struct?

Reply:
Not necessarily. The thing is that we have no idea what action block will contain, they require only two parameters - "command" and "timer".  This ties back to "builtIn" functionality that receives raw action block and extract whatever it needs to configure itself. For example "format" parameter is irrelevant for "sh" blocks but it is relevant for CPU and Memory block. In theory you could add waaaay more to action blocks and have corresponding built in functionality that makes use of it. Actually, you can extend like that without much of in-code changes. Just put whatever you need in action JSON block and extract it in your builtIn code (through action["whatever"]). As far as I remember this is exactly how "format" parameter was introduced. Somebody added 1 line in goblocks.json, and 1 line in CPU and memory blocks that used it to format themselves. 

I guess you could make a struct like:
```
type configStruct struct {
	Command string
        Timer string
	Rest []map[string]interface{}
}
```
But that doesn't get rid of []map[string]interface{}. Moreover, you are introducing extra code to essentially get rid of 2 casts. However, there might be some Go specific technique that I'm not aware of 😅.  

tl;dr: There is no struct because action blocks actually do not have structure.

Reply:
There is: `nil`.  If a value's not there, it'll be nil. Put everything into the struct, use it as needed. If you want safety, check `value != nil` before proceeding. Although your different block types _shouldn't_ be accessing fields they know will be nil for them.

On getting rid of two casts, it's worth it. Type safety is one of the reasons to use Go - bugs are reported early. What if the type you plonk in becomes a `float64` (because of a bug), and you  try to find the underlying `string` - boom!  A `panic`!

Reply:
> Put everything into the struct, use it as needed.
>
I didn't get it. Could you provide me with an example of struct that has variable number of attributes?

Reply:
```
type Action struct {
	Command        string
	UpdateSignal   int
	Timer          string
	Prefix, Suffix string
	Format         string
}

```

json.Unmarshal will leave unassigned ones at nil

Reply:
Side note: could you make updateSignal optional? Sucks to have to put it in each entry.

Side note 2: I removed the builtins in my fork. I prefer to use the commands if they're available. Consistency

Reply:
> ```
> type Action struct {
> 	Command        string
> 	UpdateSignal   int
> 	Timer          string
> 	Prefix, Suffix string
> 	Format         string
> }
> ```
> 
> json.Unmarshal will leave unassigned ones at nil

I think you missed the point. This does not address the flexibility that I was talking about.

> Side note: could you make updateSignal optional? Sucks to have to put it in each entry.
>
Yeah, that could be done. Create an issue with "enhancement" tag. I tend to forget things.

Reply:
Flexibility? You don't want flexibility, you want type-safety. (this isn't python 😉). If they're not configured, nil, there you go. What else do you want?

Reply:
> Flexibility? You don't want flexibility, you want type-safety. (this isn't python 😉).
>
I find this statement really silly. Since when programming language dictates architecture? It's a property that stems from design. You can have flexible software in any language Go/COBOL/Python/C etc.
> What else do you want?
>
I've said this before so let me reiterate. I want to be able to access **variable** number of attributes. 

You added "Format" because you know right now that format is used. So if I add a new weather block (wttr.in kinda sucks), should I add properties such as location/temperature scale/etc. to action struct? Okay and then somebody adds a block that reads S&P 500 prices, I guess him/her should add his/her properties to action block too. You want a block that shows current number of COVID-19 cases in your country? Add all your properties to Action struct. Then we end with Action struct that potentially has 100 attributes and for every block it tries to map 95% properties that do not exist. If that is not bloat than I don't know what is.

Reply:
> bloat

Well, `nil` takes up no memory. And if you think adding those fields to the struct is bloat, just don't add the functionality. Adding them to a map would be "bloat" - just uncharted bloat. Do you really want to `ok`-check everything, and cast the types? That's gonna be buggy.

Also, here's an idea - compose the structs:

```
type Block struct {
timer string
prefix, suffix string
}

type CommandBlock struct {
Block
command string
}


type WeatherBlock {
Block
location string
format string
}
```

(Adding those is bloat imo, but your call)

Reply:
> Well, nil takes up no memory.
>
I didn't research it but based on how it works in other programming languages I'm 99% sure that this statement is just false.

Even when you have a nil value you still need a reference (that points to nothing) but that reference is still a "restricted pointer" that has a size that is equal to word length of particular computer architecture that you are working on.

> Adding them to a map would be "bloat" - just uncharted bloat.
> 

Let's assume that you have a struct as following:
```
type actionStruct struct {
	Command string
        Timer string
        Sth1 string
        Sth2 string
}
```
There are 10 blocks and Sth1, Sth2 are relevant for only one respective block. In that scenario we have 10 copies of Sth1 and 10 copies of Sth2. In total we have 20 references where 18 of them point to nothing and 2 of them are pointing to something relevant. On 64 bit system that would be 144B of wasted space.

This does not occur with a map. Action blocks that do not have Sth1 and Sth2 just don't have those entries. 0B wasted space.

It is also worth to note that Go JSON parser will look for Sth1, Sth2 and that's just a waste of computation, because we know beforehand that they do not occur in 90% of entries.

>  Do you really want to ok-check everything, and cast the types? That's gonna be buggy.
> 

Interpretation of interface map is responsibility of builtIn programs not goblocks. Goblocks just takes whatever was in JSON file, gets minimal subset that it needs ( command etc.) and ships rest to appropriate program. As to ok checking command and prefix/suffix, that will be changed.

Reply:
That's why you stick to explicit, fixed data structures. This is hardly a problem for my fork, as I am staying with oh-so-flexible commands rather than bolting on loads of features. I honestly don't see the need for the builtins, but if you do desire them keep some part in a struct and add a dictionary. It'll be much easier to understand and read.

PS: 144B of wasted space? We're programmers, we make things, we read, we don't obesses about tiny bits of memory 😉. That's why you're not writing this program in assembly!